### PR TITLE
[mono] Remove MonoListItem from Corelib

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Hopefully this doesn't break anything, but admittedly the array handling here is kind of ugly.

### CODE REVIEW

File: src/mono/mono/metadata/object-internals.h [hunk 1]
```diff
@@ -358,7 +358,7 @@ struct _MonoException {
 	MonoString *remote_stack_trace;
 	gint32	    remote_stack_index;
 	/* Dynamic methods referenced by the stack trace */
-	MonoObject *dynamic_methods;
+	MonoArray *dynamic_methods;
```
Comment (lambdageek) [inline src/mono/mono/metadata/object-internals.h:361]: should still be MonoObject for mono/mono

File: src/mono/mono/mini/mini-exceptions.c [hunk 2]
```diff
@@ -2184,12 +2184,45 @@ setup_stack_trace (MonoException *mono_ex, GSList **dynamic_methods, GList *trac
...
 		if (*dynamic_methods) {
 			/* These methods could go away anytime, so save a reference to them in the exception object */
+			MonoDomain *domain = mono_domain_get ();
+#ifdef ENABLE_NETCORE
+			int methods_len = g_slist_length (*dynamic_methods);
+			MonoArray *old_methods = mono_ex->dynamic_methods;
+			int old_methods_len = 0;
+
+			if (old_methods) {
+				old_methods_len = old_methods->max_length;
```
Comment (lambdageek) [inline src/mono/mono/mini/mini-exceptions.c:2194]:
```suggestion
				old_methods_len = mono_array_length_internal (old_methods);
```

**Review by lambdageek — APPROVED** [review]

**Comment by vargaz** [comment]
Why is this needed ?

**Comment by CoffeeFlux** [comment]
Size reduction, mostly. A bunch of the native mlist stuff was being held, and it is marginally shrinking corelib.

**Comment by vargaz** [comment]
MonoListItem is just a simple class with 2 members and no methods, so this is not going to save much.

**Comment by CoffeeFlux** [comment]
It holds the mono-mlist stuff on the native side too. I doubt the savings are huge, but it was a pretty easy cleanup.

**Pull request merged**